packages/adapters/bus-adapters/resolve-bus-rabbitmq/src/index.js

```
@@ -9,70 +9,96 @@ const defaultOptions = {
-              }
-            },
-            { noAck: true }
-          )
-        )
-        .then(() => channel)
+) => {
+  try {
+    const connection = await amqp.connect(url)
+    const channel = await connection.createChannel()
+
+    await channel.assertExchange(exchange, exchangeType, {
+      durable: false
+    })
+
+    const queue = await channel.assertQueue(queueName, {
+      arguments: {
+        messageTtl: messageTtl,
+        maxLength: maxLength
+      }
+    })
+
+    await channel.bindQueue(queue.queue, exchange)
+
+    await channel.consume(
+      queueName,
+      msg => {
+        if (msg) {
+          const content = msg.content.toString()
+          const message = JSON.parse(content)
+          handler(message)
+        }
+      },
+      { noAck: true }
     )
+
+    return channel
+  } catch (e) {
+    throw new RabbitMQBusError(e)
+  }
 }
 
 function createAdapter(options) {
   let handler = () => {}
   const config = { ...defaultOptions, ...options }
-  const initPromise = init(config, event => handler(event))
+  let initPromise = null
   const { exchange, queueName, messageTtl } = config
 
   return {
-    publish: event =>
-      initPromise.then(channel => {
-        channel.publish(
-          exchange,
-          queueName,
-          new Buffer(JSON.stringify(event)),
-          // Additional options described here:
-          // http://www.squaremobius.net/amqp.node/channel_api.html#channel_publish
-          {
-            expiration: messageTtl,
-            persistent: false
-          }
-        )
-      }),
-    subscribe: callback => initPromise.then(() => (handler = callback))
+    init: async () => {
+      if (!initPromise) {
+        initPromise = init(config, event => handler(event))
+      }
+
+      try {
+        return initPromise
+      } catch (e) {
+        initPromise = null
+        throw e
+      }
+    },
+    publish: async event => {
+      const channel = await initPromise
+
+      if (!channel) {
```

It's better to throw an err, isn't it? What purpose of this code is?